modules/nlu/src/backend/tools/fastText.ts

```diff
@@ -0,0 +1,234 @@
+
+    try {
+      execFileSync(
+        FastTextWrapper.BINPATH,
+        [
+          fArgs.method,
+          '-input',
+          trainingSetPath,
+          '-output',
+          this.modelPath.replace(/\.bin$/i, ''),
+          '-lr',
+          fArgs.learningRate,
+          '-epoch',
+          fArgs.epoch,
+          '-loss',
+          'hs',
+          '-dim',
+          fArgs.dim,
+          '-bucket',
+          fArgs.bucket,
+          '-wordNgrams',
+          fArgs.wordGram,
+          '-minCount',
+          fArgs.minCount,
+          '-minn',
+          fArgs.minn,
+          '-maxn',
+          fArgs.maxn
+        ].map(x => x.toString()),
+        { stdio: 'ignore' }
+      )
+    } catch (err) {
+      throw new Error(`Error training the NLU model.
+stdout: ${err.stdout && err.stdout.toString()}
+stderr: ${err.stderr && err.stderr.toString()}
+message: ${err.message}
+status: ${err.status}
+pid: ${err.pid}
+signal: ${err.signal}
+`)
+    }
+  }
+
+  public async predict(sentence: string, k: number = DefaultFastTextQueryArgs.k): Promise<Prediction[]> {
+    const result = await FastTextWrapper._query(this.modelPath, sentence, { method: 'predict-prob', k })
+
+    const parts = result.split(/\s|\n|\r/gi).filter(x => x.trim().length)
+    const parsed: { name: string; confidence: number }[] = []
+
+    if (parts.length <= 1) {
+      return []
+    }
+
+    for (let i = 0; i < parts.length - 1; i += 2) {
+      parsed.push({
+        name: parts[i].replace(FastTextWrapper.LABEL_PREFIX, '').trim(),
+        confidence: parseFloat(parts[i + 1])
+      })
+    }
+
+    return parsed
+  }
+
+  public async wordVectors(word: string): Promise<number[]> {
+    const result = await FastTextWrapper._query(this.modelPath, word, { method: 'print-word-vectors' })
+
+    return result
+      .split(/\s|\n|\r/gi)
+      .filter(x => x.trim().length)
+      .map(x => parseFloat(x))
+      .filter(x => !isNaN(x))
+  }
+
+  private static async _query(
+    modelPath: string,
+    input: string,
+    args: Partial<FastTextQueryArgs> = DefaultFastTextQueryArgs
+  ): Promise<string> {
+    const fArgs = { ...DefaultFastTextQueryArgs, ...(args || {}) } as FastTextQueryArgs
+    const binArgs = [fArgs.method, modelPath]
+
+    if (fArgs.method === 'predict-prob') {
+      binArgs.push('-')
+      binArgs.push(fArgs.k.toString())
+    } else if (fArgs.method === 'nn') {
+      binArgs.push(fArgs.k.toString())
+    }
+
+    try {
+      const process = await this._acquireProcess(modelPath, binArgs)
+      return new Promise<string>(resolve => {
+        process.stdin.write(`${input}${EOL}`)
+        process.stdout.once('data', resolve)
+      })
+    } finally {
+      this._releaseProcess(modelPath)
+    }
+  }
+
+  private static _scheduleProcessCleanup(modelPath: string, expiry: number) {
+    setTimeout(async () => {
+      if (this.process_cache_expiry[modelPath] !== expiry) {
```

Comment:
shouldn't it be < or > ? if there's a glitch, there is a chance that it won't match exactly the time specified

Reply:
Refactored 👍